Tests/SQLiteNIOTests/SQLiteCustomFunctionTests.swift

```diff
@@ -0,0 +1,374 @@
+										.map { rows in rows[0].column("result")?.integer }.wait())
+		XCTAssertEqual(2, try conn.query("SELECT f(1, 2) as result")
+										.map { rows in rows[0].column("result")?.integer }.wait())
+		XCTAssertEqual(3, try conn.query("SELECT f(1, 1, 1) as result")
+										.map { rows in rows[0].column("result")?.integer }.wait())
+	}
+
+	// MARK: - Errors
+
+	func testFunctionThrowingDatabaseCustomErrorWithMessage() throws {
+		let conn = try SQLiteConnection.open(storage: .memory, threadPool: self.threadPool, on: self.eventLoop).wait()
+		defer { try! conn.close().wait() }
+
+		struct MyError: Error {
+			let message: String
+		}
+
+		let fn = SQLiteCustomFunction("f") { _ in
+			throw MyError(message: "custom message")
+		}
+
+		try conn.install(customFunction: fn).wait()
+
+		do {
+			_ = try conn.query("SELECT f()").wait()
+			XCTFail("Expected Error")
+		} catch let error as MyError {
+			XCTFail("expected this not to match")
+			XCTAssertEqual(error.message, "custom message")
+		} catch let error as SQLiteError {
+
+			XCTAssertEqual(error.reason, .error)
+			XCTAssertEqual(error.message, "MyError(message: \"custom message\")")
+		}
+	}
+
+	func testFunctionThrowingNSError() throws {
+		let conn = try SQLiteConnection.open(storage: .memory, threadPool: self.threadPool, on: self.eventLoop).wait()
+		defer { try! conn.close().wait() }
+		let fn = SQLiteCustomFunction("f") { _ in
+			throw NSError(domain: "CustomErrorDomain", code: 123, userInfo: [NSLocalizedDescriptionKey: "custom error message"])
+		}
+
+		try conn.install(customFunction: fn).wait()
+
+		do {
+			_ = try conn.query("SELECT f()").wait()
+			XCTFail("Expected Error")
+		} catch let error as SQLiteError {
+
+			XCTAssertEqual(error.reason, .error)
+			XCTAssertTrue(error.message.contains("CustomErrorDomain"))
+			XCTAssertTrue(error.message.contains("123"))
+			#if os(Linux)
+			XCTAssertTrue(error.message.contains("(null)"), "expected '\(error.message)' to contain '(null)'")
+			#else
+			XCTAssertTrue(error.message.contains("custom error message"), "expected '\(error.message)' to contain 'custom error message'")
+			#endif
+		}
+	}
+
+	// MARK: - Misc
+
+	func testFunctionsAreClosures() throws {
+		let conn = try SQLiteConnection.open(storage: .memory, threadPool: self.threadPool, on: self.eventLoop).wait()
+		defer { try! conn.close().wait() }
+
+		var x = 123
+		let fn = SQLiteCustomFunction("f", argumentCount: 0) { dbValues in
+			return x
+		}
+		try conn.install(customFunction: fn).wait()
+		x = 321
+		XCTAssertEqual(321, try conn.query("SELECT f() as result").map({ rows in rows[0].column("result")?.integer }).wait())
+	}
+
+	// MARK: - setup
+
+	var threadPool: NIOThreadPool!
+	var eventLoopGroup: EventLoopGroup!
+	var eventLoop: EventLoop {
+		return self.eventLoopGroup.next()
+	}
+
+	override func setUp() {
+		self.threadPool = .init(numberOfThreads: 8)
+		self.threadPool.start()
+		self.eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 8)
```

Comment:
Please use `1` as per the existing tests, or if multiple threads are strictly required, `System.coreCount`.

Reply:
Done